### PR TITLE
Replace conda with mamba

### PR DESCRIPTION
**Pull request by crusaderky**
Shave 2 minutes off each CI workflow
Blocked by #7383

### CODE REVIEW

File: .github/workflows/ci-additional.yml [hunk 1]
```diff
@@ -18,7 +18,8 @@ jobs:
       - name: Setup Conda Environment
         uses: conda-incubator/setup-miniconda@v2
         with:
-          miniconda-version: "latest"
+          mamba-version: "*"
```
Comment (xhochy) [inline .github/workflows/ci-additional.yml:21]: Probably useful to have a look at the master of the `setup-miniconda` actions and use `miniforge-variant: Mambaforge`. This gives you a base environment with mamba pre-installed and `conda-forge` as a channel.

Comment (crusaderky) [reply .github/workflows/ci-additional.yml]: Doesn't work:
```
Warning: Unexpected input(s) 'miniforge-variant', 'miniforge-version', 'use-mamba', valid inputs are ['installer-url', 'miniconda-version', 'conda-version', 'conda-build-version', 'environment-file', 'activate-environment', 'python-version', 'add-anaconda-token', 'add-pip-as-python-dependency', 'allow-softlinks', 'auto-activate-base', 'auto-update-conda', 'condarc-file', 'channel-alias', 'channel-priority', 'channels', 'show-channel-urls', 'use-only-tar-bz2', 'remove-profiles', 'mamba-version', 'architecture']
```

Comment (xhochy) [reply .github/workflows/ci-additional.yml:22]: They are only yet available on `master`, not on `v2`. If they also work here well, `setup-miniconda` could do a new release to have them available in a stable fashion.

Comment (crusaderky) [reply .github/workflows/ci-additional.yml]: (ignore; I was still using @v2)

Comment (crusaderky) [reply .github/workflows/ci-additional.yml]: it works now, thanks

**Comment by crusaderky** [comment]
There is a mamba-specific bug on MacOS Python 3.6 / numpy 1.15:
```
    try:
>       import scipy.stats
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   from .stats import *
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   from scipy.spatial.distance import cdist

../../../miniconda3/envs/test-environment/lib/python3.6/site-packages/scipy/stats/stats.py:174: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   from .qhull import *
E   ImportError: dlopen(/Users/runner/miniconda3/envs/test-environment/lib/python3.6/site-packages/scipy/spatial/qhull.cpython-36m-darwin.so, 2): Library not loaded: @rpath/libopenblas.dylib
E     Referenced from: /Users/runner/miniconda3/envs/test-environment/lib/python3.6/site-packages/scipy/spatial/qhull.cpython-36m-darwin.so
E     Reason: image not found

../../../miniconda3/envs/test-environment/lib/python3.6/site-packages/scipy/spatial/__init__.py:98: ImportError
```
I'm parking this PR for now and I'm going to retest it after dropping Python 3.6

**Comment by mrocklin** [comment]
cc @jakirkham who might find this interesting

**Comment by jakirkham** [comment]
Would suggest not using `@master` as future changes could break this workflow. If we want to use this functionality, would suggest raising an issue on that repo asking for a new tag.

**Comment by xhochy** [comment]
Raised this in https://github.com/conda-incubator/setup-miniconda/issues/106#issuecomment-780800170

File: .github/workflows/ci-additional.yml [hunk 2]
```diff
@@ -16,9 +16,11 @@ jobs:
         uses: actions/checkout@v2
 
       - name: Setup Conda Environment
-        uses: conda-incubator/setup-miniconda@v2
+        uses: conda-incubator/setup-miniconda@master
```
Comment (thomasjpfan) [inline .github/workflows/ci-additional.yml:19]: If you want this now, you can pin this to a commit:

```suggestion
        uses: conda-incubator/setup-miniconda@77b16ed746da28724c61e1f1ad23395a4b695ef5
```

and then make a follow up PR when a new version gets tagged.

**Comment by crusaderky** [comment]
xref https://github.com/conda-incubator/setup-miniconda/issues/106

**Comment by crusaderky** [comment]
mamba is broken on numpy 1.15 regardless of Python version; see above.

**Comment by crusaderky** [comment]
I've split out the bump to numpy 1.16 into a separate PR, which is now a dependency of this one: #7383

File: continuous_integration/scripts/install.sh [hunk 3]
```diff
@@ -7,9 +7,12 @@ set -xe
 # python -m pip install --no-deps cityhash
 
 if [[ ${UPSTREAM_DEV} ]]; then
-    conda update -y -c arrow-nightlies pyarrow
+    mamba update -y -c arrow-nightlies pyarrow
 
+    # FIXME https://github.com/mamba-org/mamba/issues/412
+    # mamba uninstall --force numpy pandas
     conda uninstall --force numpy pandas
```
Comment (jakirkham) [inline continuous_integration/scripts/install.sh:14]: xref: https://github.com/mamba-org/mamba/issues/412

**Comment by crusaderky** [comment]
@jrbourbeau ready for merge. Weirdly, unlike in dask/distributed I don't need to specify the channels in the workflow.

**Review by jrbourbeau — APPROVED** [review]

**Comment by wolfv** [comment]
@crusaderky cool to see this! Any info on what is broken with mamba / numpy 1.15? is it mamba or broken metadata or numpy? If there is something we can fix in mamba, please let us know!

**Comment by crusaderky** [comment]
@wolfv https://github.com/dask/dask/runs/2101922250

```
>   import scipy.stats
>   from .stats import *
>   from scipy.spatial.distance import cdist
>   from .qhull import *
E   ImportError: dlopen(/Users/runner/miniconda3/envs/test-environment/lib/python3.7/site-packages/scipy/spatial/qhull.cpython-37m-darwin.so, 2): Library not loaded: @rpath/libopenblas.dylib
E     Referenced from: /Users/runner/miniconda3/envs/test-environment/lib/python3.7/site-packages/scipy/spatial/qhull.cpython-37m-darwin.so
E     Reason: image not found
```
this happens on MacOSX only. conda list output is linked above.

**Comment by wolfv** [comment]
Ok, looks like mixing defaults & conda-forge channels.
Any reason for keeping defaults in there?

Otherwise, at least using strict-channel priority should help (as recommended by conda-forge).

**Comment by crusaderky** [comment]
@wolfv the exact same environment file works fine with conda.
defaults are there as a fallback for the cases when there isn't a package in conda-forge

**Pull request merged**